spec/models/user_spec.rb

```diff
@@ -0,0 +1,106 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  let!(:password) { SecureRandom.uuid }
+  let(:user_attributes) {
+    {
+      email: "foo@example.com",
+      password: password,
+      password_confirmation: password
+    }
+  }
+  describe "#save" do
+    let(:user) { described_class.new(user_attributes) }
+    describe "api key generation" do
+      context "given a valid user" do
+        it "creates a uuid for the api key" do
+          expect(user.api_key).not_to be_present
+          user.save
+          expect(user.api_key).to be_present
+        end
+      end
+    end
+  end
+
+  describe "#authenticate" do
+    let(:user) { described_class.new(user_attributes) }
+    context "given the correct secret key" do
+      it "returns the user" do
+        expect(user.authenticate(password)).to eq(user)
+      end
+    end
+
+    context "given the incorrect secret_key" do
+      it "returns false" do
+        expect(user.authenticate("not the secret key")).to be_falsey
+      end
+    end
+  end
+
+  describe "#valid?" do
+    let(:user) { described_class.new(user_attributes) }
+    context "given the presence of an email and password" do
+      it "is valid" do
+        expect(user).to be_valid
+      end
+    end
+
+    context "given the absence of an email or password" do
+      let(:user) { described_class.new() }
+      it "is invalid" do
+        expect(user).not_to be_valid
+        expect(user.errors[:email]).to be_present
+        expect(user.errors[:password]).to be_present
+      end
+    end
+
+    context "given a pre-existing email" do
+      let(:user) { described_class.new(user_attributes) }
+
+      before do
+        described_class.create!(user_attributes)
+      end
+
+      it "is invalid" do
+        expect(user).not_to be_valid
+        expect(user.errors[:email]).to be_present
+        expect(user.errors.details[:email]).to match(a_collection_including(a_hash_including(error: :taken)))
+      end
+    end
+  end
+
+  describe ".from_token_request" do
```

Comment:
Interesting.  Thanks for the link.

Let's follow the convention they've outlined here.  I'd rather be able to point future devs to that doc than have to explain our own set of conventions.